quil-rs/src/instruction/extern_call.rs

```diff
@@ -0,0 +1,2060 @@
+    /// Create a new call instruction with resolved arguments. This will validate the
+    /// name as a user identifier.
+    pub fn try_new(
+        name: String,
+        arguments: Vec<UnresolvedCallArgument>,
+    ) -> Result<Self, CallValidationError> {
+        validate_user_identifier(name.as_str()).map_err(CallValidationError::Name)?;
+
+        Ok(Self {
+            name,
+            arguments: CallArguments::Unresolved(arguments),
+        })
+    }
+}
+
+/// An error that can occur when resolving a call instruction.
+#[derive(Clone, Debug, thiserror::Error, PartialEq)]
+pub enum CallArgumentError {
+    /// The return argument could not be resolved.
+    #[error("error resolving return argument: {0:?}")]
+    Return(CallArgumentResolutionError),
+    /// An argument could not be resolved.
+    #[error("error resolving argument {index}: {error:?}")]
+    Argument {
+        index: usize,
+        error: CallArgumentResolutionError,
+    },
+}
+
+/// An error that can occur when resolving a call instruction to a specific
+/// [`ExternSignature`].
+#[derive(Debug, thiserror::Error, PartialEq)]
+pub enum CallSignatureError {
+    #[error("expected {expected} arguments, found {found}")]
+    ParameterCount { expected: usize, found: usize },
+    #[error("error resolving arguments: {0:?}")]
+    Arguments(Vec<CallArgumentError>),
+}
+
+/// An error that can occur when resolving a call instruction within the context
+/// of a program.
+#[derive(Debug, thiserror::Error, PartialEq)]
+pub enum CallResolutionError {
+    /// A matching extern instruction was found, but the signature validation failed.
+    #[error("call found matching extern instruction for {name}, but signature validation failed: {error:?}")]
+    Signature {
+        name: String,
+        error: CallSignatureError,
+    },
+    /// A matching extern instruction was found, but it has no signature.
+    #[error("call found matching extern instruction for {0}, but it has no signature")]
+    NoSignature(String),
+    /// No matching extern instruction was found.
+    #[error("no extern instruction found with name {0}")]
+    NoMatchingExternInstruction(String),
+}
+
+#[allow(clippy::manual_try_fold)]
+fn convert_unresolved_to_resolved_call_arguments(
+    arguments: &[UnresolvedCallArgument],
+    signature: &ExternSignature,
+    memory_regions: &IndexMap<String, MemoryRegion>,
+) -> Result<Vec<ResolvedCallArgument>, CallSignatureError> {
+    arguments
+        .iter()
+        .enumerate()
+        .map(|(i, argument)| {
+            if i == 0 {
+                if let Some(return_type) = signature.return_type {
+                    return argument
+                        .resolve_return(memory_regions, return_type)
+                        .map_err(CallArgumentError::Return);
+                }
+            }
+            let parameter_index = if signature.return_type.is_some() {
+                i - 1
+            } else {
+                i
+            };
```

Comment:
I find this indexing logic confusing.  Perhaps we could reflect it better either in the type or in another function that bundles up this logic?